lua/mini/misc.lua

```diff
@@ -409,6 +409,139 @@ MiniMisc.zoom = function(buf_id, config)
+  end
+
+  if not H.is_boolean(opts.center) then
+    H.error('In `setup_restore_cursor()` `opts.center` should be a boolean.')
+  end
+
+  -- local augroup = vim.api.nvim_create_augroup('MiniMiscRestoreCursor',
+  --                                             { clear = true })
+  -- local au_command = string.format("lua require('mini.misc').restore_cursor_buftype(%s)",
+  --                                  vim.inspect(opts, { newline = ' ', indent = '' }))
+  -- vim.api.nvim_create_autocmd(
+  --   'BufReadPre',
+  --   {
+  --     group = augroup,
+  --     command = au_command,
+  --     desc = 'Setup the restore cursor autocmd',
+  --   })
+
+  local au_command = string.format(
+    [[augroup MiniMiscRestoreCursor
+      au!
+      au BufReadPre * au FileType <buffer> ++once lua require('mini.misc').restore_cursor(%s)
+    augroup END]],
+    vim.inspect(opts, { newline = ' ', indent = '' }))
+  vim.api.nvim_exec(au_command, false)
+end
+
+-- MiniMisc.restore_cursor_buftype = function(opts)
+--   -- Check if the buffer should be ignored
+--   if vim.tbl_contains(opts.ignore_buftype,
+--                       vim.api.nvim_buf_get_option(0, "buftype")) then
+--     return
+--   end
+--
+--   local augroup = vim.api.nvim_create_augroup('MiniMiscRestoreCursor',
+--                                               { clear = false })
+--   local au_command = string.format("lua require('mini.misc').restore_cursor(%s)",
+--                                    vim.inspect(opts, { newline = ' ', indent = '' }))
+--   vim.api.nvim_create_autocmd(
+--     'FileType',
+--     {
+--       group = augroup,
+--       command = au_command,
+--       buffer = vim.api.nvim_get_current_buf(),
+--       once = true,
+--       desc = 'Restore the cursor position',
+--     })
+-- end
+
+--- Restore cursor when reopening a file
+MiniMisc.restore_cursor = function(opts)
+  -- Check if the buffer should be ignored
+  if vim.tbl_contains(opts.ignore_buftype,
+                      vim.api.nvim_buf_get_option(0, "buftype")) then
+    return
+  end
+
+  -- Check if the filetype should be ignored
+  if vim.tbl_contains(opts.ignore_filetype,
+                      vim.api.nvim_buf_get_option(0, "filetype")) then
+      -- reset cursor to first line
+      vim.api.nvim_command([[normal! gg]])
+      return
+  end
+
+  -- Place cursor --
+
+  -- If a line has already been specified on the command line, we are done
+  --   nvim file +num
+  if vim.fn.line(".") > 1 then
+      return
+  end
+
+  local last_line = vim.fn.line([['"]])
+  local buff_last_line = vim.fn.line("$")
+
+  if last_line > 0 and last_line <= buff_last_line then
+    if opts.center then
+      vim.api.nvim_command([[normal! g`"zz]])
+    else
+      vim.api.nvim_command([[normal! g`"]])
+    end
+  end
+
+  -- Open fold
+  if vim.fn.foldclosed(".") ~= -1 then
+    vim.api.nvim_command([[normal! zvzz]])
+  end
+end
```

Comment:
Maybe all these can be replaced with this?

```lua
vim.cmd('normal! g`"zv')
if opt.center then vim.cmd('normal! zz') end
```

Doing `zv` inside a non-fold seems harmless from experience.

Also, do you have any bad experience with using `` g`" `` instead of `g'"`? I quickly tried couple of days ago the first one and it sometimes gave some weird behavior of off-by-one placement.

Reply:
> Maybe all these can be replaced with this?
> 
> ```lua
> vim.cmd('normal! g`"zv')
> if opt.center then vim.cmd('normal! zz') end
> ```
> 
> Doing `zv` inside a non-fold seems harmless from experience.

I just tried and sometimes the cursor is placed somewhere else if you don't have any folds. However I made the code a bit nicer.

> Also, do you have any bad experience with using `` g`" `` instead of `g'"`? I quickly tried couple of days ago the first one and it sometimes gave some weird behavior of off-by-one placement.

I don't have any problems with ``g`"``, it works just fine.

Reply:
> I just tried and sometimes the cursor is placed somewhere else if you don't have any folds. However I made the code a bit nicer.

Hmm... I added these lines and tried to open several files:

```lua
_G.restore_cursor = function() vim.cmd([[normal! g`"zvzz]]) end
vim.cmd([[autocmd BufRead * lua _G.restore_cursor()]])
```

Didn't have any issues with folds. Could you isolate the problem and make it reproducible? It is not only for this module, but in case there is an issue I should recheck other modules.

> I don't have any problems with ``g`"``, it works just fine.

Had one issue when opening file with 'telescope.nvim': it places cursor one column to the left where it is needed... Strange. But if it is only in Telescope, then no big deal.

Reply:
> Had one issue when opening file with 'telescope.nvim': it places cursor one column to the left where it is needed... Strange. But if it is only in Telescope, then no big deal.

I can reproduce it with telescope. However using ``g'"`` doesn't fix it for me. It always moves the cursor. Might be a telescope bug.